engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,567 @@ package ingestion
+	} else {
+		sb.WriteString("[manual]")
+	}
+
+	return sb.String()
 }
 
-type StopControlState byte
-
-const (
-	// StopControlOff default state, envisioned to be used most of the time.
-	// Stopping module is simply off, blocks will be processed "as usual".
-	StopControlOff StopControlState = iota
-
-	// StopControlSet means stopHeight is set but not reached yet,
-	// and nothing related to stopping happened yet.
-	// We could still go back to StopControlOff or progress to StopControlCommenced.
-	StopControlSet
-
-	// StopControlCommenced indicates that stopping process has commenced
-	// and no parameters can be changed anymore.
-	// For example, blocks at or above stopHeight has been received,
-	// but finalization didn't reach stopHeight yet.
-	// It can only progress to StopControlPaused
-	StopControlCommenced
-
-	// StopControlPaused means EN has stopped processing blocks.
-	// It can happen by reaching the set stopping `stopHeight`, or
-	// if the node was started in pause mode.
-	// It is a final state and cannot be changed
-	StopControlPaused
-)
+type StopControlOption func(*StopControl)
+
+// StopControlWithLogger sets logger for the StopControl
+// and adds a "component" field to it
+func StopControlWithLogger(log zerolog.Logger) StopControlOption {
+	return func(s *StopControl) {
+		s.log = log.With().Str("component", "stop_control").Logger()
+	}
+}
+
+func StopControlWithStopped() StopControlOption {
+	return func(s *StopControl) {
+		s.stopped = true
+	}
+}
+
+func StopControlWithVersionControl(
+	nodeVersion *semver.Version,
+	versionBeacons storage.VersionBeacons,
+	crashOnVersionBoundaryReached bool,
+) StopControlOption {
+	return func(s *StopControl) {
+		s.nodeVersion = nodeVersion
+		s.versionBeacons = versionBeacons
+		s.crashOnVersionBoundaryReached = crashOnVersionBoundaryReached
+	}
+}
+
+// StopControlHeaders is an interface for fetching headers
+// Its jut a small subset of storage.Headers for comments see storage.Headers
+type StopControlHeaders interface {
+	ByHeight(height uint64) (*flow.Header, error)
+}
 
 // NewStopControl creates new empty NewStopControl
 func NewStopControl(
-	log zerolog.Logger,
-	paused bool,
-	lastExecutedHeight uint64,
+	headers StopControlHeaders,
+	options ...StopControlOption,
 ) *StopControl {
-	state := StopControlOff
-	if paused {
-		state = StopControlPaused
+
+	sc := &StopControl{
+		log:     zerolog.Nop(),
+		headers: headers,
 	}
-	log.Debug().Msgf("created StopControl module with paused = %t", paused)
-	return &StopControl{
-		log:                    log,
-		state:                  state,
-		highestExecutingHeight: lastExecutedHeight,
+
+	for _, option := range options {
+		option(sc)
 	}
-}
 
-// GetState returns current state of StopControl module
-func (s *StopControl) GetState() StopControlState {
-	s.RLock()
-	defer s.RUnlock()
-	return s.state
+	log := sc.log.With().
+		Bool("node_will_react_to_version_beacon",
+			sc.nodeVersion != nil).
```

Comment:
> Maybe, independently of the version beacon, we want to say that we won't start up the node if it does not have a parseable software version. But I don't know why we would want to disable the version beacon in this case.

Yes this is what we should do, but that requires a bit more work around how we tag our versions (git hook or CI or something). Until we have that in place I'm not sure it makes sense to enforce that the nodes have a semver compliant version on startup. However it is a good point that an invalid results might be produced if we are not careful.

I think it might be best to add the check that on startup the node version should be semver compliant (+ the "v") in a separate smaller PR so that its easier to point people to it and let everyone know.